### PR TITLE
Review opset1 unsqueeze for shape inference aspects

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Review shape inference for shape, label, partial values propagation (add tests)
 - Add template `shape_infer` implementation with tests
 - Fix segfault when repeated axes used

### Tickets:
 - 90598
 - 90695
 - 90732
 - 90769
 - 90806
 - 90843
 - 90880
 - 90917

### CODE REVIEW

File: src/core/tests/type_prop/unsqueeze.cpp [hunk 1]
```diff
@@ -62,35 +80,193 @@ TEST(type_prop, unsqueeze_empty_axes) {
...
 
-    auto bc = std::make_shared<op::v1::Broadcast>(param, unsqueeze);
-    ASSERT_EQ(bc->get_shape(), (Shape{3}));
+    const auto bc = std::make_shared<op::v1::Broadcast>(param, unsqueeze);
 
-    const auto& output_shape = bc->get_output_partial_shape(0);
-    ASSERT_EQ(ov::DimensionTracker::get_label(output_shape[0]), 10);
+    EXPECT_EQ(bc->get_output_partial_shape(0), exp_shape);
+    const auto labels = get_shape_labels(bc->get_output_partial_shape(0));
+    EXPECT_THAT(labels, ElementsAre(exp_labels.front()));
```
Comment (mitruska) [inline src/core/tests/type_prop/unsqueeze.cpp:271]: It's a little bit hard to follow the logic and shapes used in this test, and what the expected labels really are in this case.
Additional test case with single `Unsqueeze` and example of input shape and expected labels could be added.

File: src/core/shape_inference/include/unsqueeze_shape_inference.hpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    std::vector<int64_t> axes_val;
+    const auto has_axes = get_data_as_int64<T>(1, op, axes_val, constant_data);
+
+    if (has_axes && arg_shape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op, !axes_val.empty(), "'axes' input is mandatory");
+        // Remove repeated axes on input
+        std::unordered_set<int64_t> tmp(std::make_move_iterator(axes_val.begin()),
+                                        std::make_move_iterator(axes_val.end()));
+        std::vector<int64_t> unique_axes(std::make_move_iterator(tmp.begin()), std::make_move_iterator(tmp.end()));
```
Comment (mitruska) [inline src/core/shape_inference/include/unsqueeze_shape_inference.hpp:30]: As we discussed, no need to use `move` for primitive types.

File: src/plugins/intel_cpu/tests/unit/shape_inference_test/unsqueeze_shape_inference_test.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                                make_tuple(ShapeVector{{2}, {1}}, std::vector<int64_t>{1}, StaticShape({2, 1})),
+                                make_tuple(ShapeVector{{2}, {1}}, std::vector<int64_t>{-1}, StaticShape({2, 1})),
+                                make_tuple(ShapeVector{{2}, {1}}, std::vector<int64_t>{-2}, StaticShape({1, 2}))),
+                         PrintToStringParamName());
+
+INSTANTIATE_TEST_SUITE_P(
+    multi_dim_shapes,
+    UnsqueezeStaticShapeInferenceTest,
+    Values(make_tuple(ShapeVector{{2, 3}, {2}}, std::vector<int64_t>{0, 3}, StaticShape({1, 2, 3, 1})),
+           make_tuple(ShapeVector{{3, 2}, {3}}, std::vector<int64_t>{0, 2, 4}, StaticShape({1, 3, 1, 2, 1})),
```
Comment (mitruska) [inline src/plugins/intel_cpu/tests/unit/shape_inference_test/unsqueeze_shape_inference_test.cpp:89]: It would be good to add similar test case with unsorted axes.

File: src/core/src/op/unsqueeze.cpp [hunk 4]
```diff
@@ -82,29 +62,30 @@ bool evaluate(const HostTensorPtr& arg0, const HostTensorPtr& out) {
...
-    });
     // Sort in increasing order
-    std::set<int64_t, less<int64_t>> axes_set(axes.begin(), axes.end());
+    std::set<int64_t> axes_set(std::make_move_iterator(axes.begin()), std::make_move_iterator(axes.end()));
     NGRAPH_CHECK(axes.size() == axes_set.size(), "Axes has duplicate axis.");
+
+    auto out_shape = data_shape;
     for (int64_t axis : axes_set) {
-        NGRAPH_CHECK(axis >= 0 && axis < out_rank, "Axis is out of bounds: ", axis);
         out_shape.insert(out_shape.begin() + axis, 1);
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/op/unsqueeze.cpp:89]: Shape infer and evaluate are not calculating output shape in same way and also use different validation.
Maybe in next version of operator it can be unified if shape infer follow `numpy` and reject repeated axes, then this shape calculations should be same.

File: src/core/tests/type_prop/unsqueeze.cpp (hunk 1 above)
Comment (praasz) [reply src/core/tests/type_prop/unsqueeze.cpp]: It expect that labels got one element which got value of first from input (expected) labels, no sure what is not clear.

Added tests for label propagation for single `Unsqueeze`

**Review by mitruska — APPROVED** [review]

File: src/core/src/op/unsqueeze.cpp [hunk 5]
```diff
@@ -82,29 +62,30 @@ bool evaluate(const HostTensorPtr& arg0, const HostTensorPtr& out) {
...
     out->set_element_type(element_type);
 
-    auto data_shape = arg0->get_shape();
-    int64_t data_rank = static_cast<int64_t>(data_shape.size());
-    auto axes_shape = arg1->get_shape();
-    NGRAPH_CHECK(axes_shape.size() == 1 || axes_shape.empty(),
-                 "Axes to add must be a scalar or 1D tensor with 1 element");
+    const auto& axes_shape = arg1->get_shape();
+    OPENVINO_ASSERT(ov::is_scalar(axes_shape) || ov::is_vector(axes_shape),
+                    "Axes to add must be a scalar or 1D tensor with 1 element");
```
Comment (jane-intel) [inline src/core/src/op/unsqueeze.cpp:74]: we could skip it if we incorporate the same check in the shape_infer

File: src/core/src/op/unsqueeze.cpp [hunk 6]
```diff
@@ -82,29 +62,30 @@ bool evaluate(const HostTensorPtr& arg0, const HostTensorPtr& out) {
...
+
+    const auto& data_shape = arg0->get_shape();
+    const auto out_rank = static_cast<int64_t>(data_shape.size() + shape_size(axes_shape));
+
+    // Get axes and normalize
+    auto axes = read_index_vector(arg1);
+    normalize_axes(node, out_rank, axes);
```
Comment (jane-intel) [inline src/core/src/op/unsqueeze.cpp:81]: why can't we call shape_infer with axis passed as HostTensor?

File: src/core/src/op/unsqueeze.cpp [hunk 7]
```diff
@@ -27,38 +24,21 @@ op::v0::Unsqueeze::Unsqueeze(const Output<Node>& data, const Output<Node>& axes)
...
-    const auto data_rank = data_partial_shape.rank();
 
-    const auto axes_constant = get_constant_from_source(input_value(1));
-    auto axes_pshape = get_input_partial_shape(1);
+    const auto& axes_pshape = get_input_partial_shape(1);
 
     NODE_VALIDATION_CHECK(this,
                           axes_pshape.rank().compatible(0) || axes_pshape.rank().compatible(1),
                           "Second input (axes) should not be of rank higher than 1. Got: ",
                           axes_pshape.rank().get_length());
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/op/unsqueeze.cpp:33]: Will move it to `shape_infer` and unify with check done in `evaluate` mthod.

File: src/core/src/op/unsqueeze.cpp [hunk 8]
```diff
@@ -27,38 +24,21 @@ op::v0::Unsqueeze::Unsqueeze(const Output<Node>& data, const Output<Node>& axes)
...
-    set<int64_t> axes(begin(normalized_axes), end(normalized_axes));
-    vector<Dimension> output_shape{data_partial_shape};
-    for (auto axis : axes) {
-        NODE_VALIDATION_CHECK(this, axis <= expanded_rank, "provided 'axes' value ", axis, " is not valid.");
+    shape_infer(this, input_shapes, output_shapes);
 
-        output_shape.insert(next(begin(output_shape), axis), 1);
-    }
-    set_output_type(0, get_input_element_type(0), ov::PartialShape{output_shape});
+    set_output_size(output_shapes.size());
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/op/unsqueeze.cpp:40]: Removed.

File: src/core/src/op/unsqueeze.cpp (hunk 5 above)
Comment (praasz) [reply src/core/src/op/unsqueeze.cpp:74]: Check unified with shape_infer but still there. As `evaluate` calculates output shapes different cannot use `shape_infer` here now.

Added comment about the reason and TODO for next version of operator.

File: src/core/src/op/unsqueeze.cpp (hunk 6 above)
Comment (praasz) [reply src/core/src/op/unsqueeze.cpp:81]: When shape inference use same as numpy checks for repeated axis then evaluate can use it added comment about it.

**Review by jane-intel — APPROVED** [review]
